--- authentication_context.orig	2016-11-03 14:59:47.165017595 +0800
+++ authentication_context.py	2016-11-03 15:27:08.065519586 +0800
@@ -124,6 +124,28 @@
 
         return self._acquire_token(token_func)
 
+    #
+    #   Added by Charles.Hsueh
+    #
+    def acquire_token_with_username_password(self, resource, username, password, client_id, client_secret):
+        '''Gets a token for a given resource via user credentails.
+
+        :param str resource: A URI that identifies the resource for which the
+            token is valid.
+        :param str username: The username of the user on behalf this
+            application is authenticating.
+        :param str password: The password of the user named in the username
+            parameter.
+        :param str client_id: The OAuth client id of the calling application.
+        :returns: dict with several keys, include "accessToken" and
+            "refreshToken".
+        '''
+        def token_func(self):
+            token_request = TokenRequest(self._call_context, self, client_id, resource)
+            return token_request.get_token_with_username_password(username, password, client_secret)
+
+        return self._acquire_token(token_func)
+
     def acquire_token_with_client_credentials(self, resource, client_id, client_secret):
         '''Gets a token for a given resource via client credentials.
 
